libraries/AP_Scheduler/AP_Scheduler.tpp

```diff
@@ -0,0 +1,213 @@
+    // @Values: 0:Disabled,2:ShowSlips,3:ShowOverruns
+    // @User: Advanced
+    AP_GROUPINFO("DEBUG",    0, AP_Scheduler, _debug, 0),
+
+    // @Param: LOOP_RATE
+    // @DisplayName: Scheduling main loop rate
+    // @Description: This controls the rate of the main control loop in Hz. This should only be changed by developers. This only takes effect on restart
+    // @Values: 50:50Hz,100:100Hz,200:200Hz,250:250Hz,300:300Hz,400:400Hz
+    // @RebootRequired: True
+    // @User: Advanced
+    AP_GROUPINFO("LOOP_RATE",  1, AP_Scheduler, _loop_rate_hz, 50),
+
+    AP_GROUPEND
+};
+
+// constructor
+template <class Vehicle>
+AP_Scheduler<Vehicle>::AP_Scheduler(Vehicle *parent, const uint16_t loop_rate) : _parent(parent)
+{
+    _tasks = nullptr;
+    _last_run = nullptr;
+  
+    AP_Param::setup_object_defaults(this, var_info);
+    _loop_rate_hz.set_and_save(loop_rate);
+
+    // only allow 50 to 400 Hz
+    if (_loop_rate_hz < 50) {
+        _loop_rate_hz.set(50);
+    } else if (_loop_rate_hz > 400) {
+        _loop_rate_hz.set(400);
+    }
+}
+
+// initialise the scheduler
+template <class Vehicle>
+AP_Scheduler<Vehicle>::~AP_Scheduler() {
+    if(_last_run) {
+        delete [] _last_run;
+    }
+    if(_perf_counters) {
+        delete [] _perf_counters;
+    }
+}
+
+// initialise the scheduler
+template <class Vehicle>
+void AP_Scheduler<Vehicle>::init(const AP_Task<Vehicle> *tasks, uint8_t num_tasks)
+{
+    if(tasks == nullptr || num_tasks == 0) {
+        return;
+    }
+    _tasks = tasks;
+    _num_tasks = num_tasks;
+    _last_run = new uint16_t[_num_tasks];
+    memset(_last_run, 0, sizeof(_last_run[0]) * _num_tasks);
+    _tick_counter = 0;
+}
+
+// one tick has passed
+template <class Vehicle>
+void AP_Scheduler<Vehicle>::tick(void)
+{
+    _tick_counter++;
+}
+
+/*
+  run one tick
+  this will run as many scheduler tasks as we can in the specified time
+ */
+template <class Vehicle>
+void AP_Scheduler<Vehicle>::run(uint32_t time_available)
+{
+    // break condition
+    if(_tasks == nullptr) {
+        return;
+    }
+  
+    if (_debug > 3 && _perf_counters == nullptr) {
+        _perf_counters = new AP_HAL::Util::perf_counter_t[_num_tasks];
+        for (uint8_t i=0; i<_num_tasks && _perf_counters != nullptr; i++) {
+            _perf_counters[i] = hal.util->perf_alloc(AP_HAL::Util::PC_ELAPSED, _tasks[i]._name);
+        }
+    }
+
+    for (uint8_t i=0; i<_num_tasks && _tasks[i]._max_time_micros <= time_available; i++) {
+        const uint16_t dt = _tick_counter - _last_run[i];
+        const uint16_t interval_ticks = (_loop_rate_hz / _tasks[i]._rate_hz) < 1 ? 1 : (_loop_rate_hz / _tasks[i]._rate_hz);
+                
+        if(dt < interval_ticks) {
+            continue;
+        }
+    
```

Comment:
here there's a missing check for "task time allowed" < "time available", and just continue the loop in case this doesn't hold true.